src/rdkafka_ssl.c

```diff
@@ -417,6 +417,123 @@ rd_kafka_transport_ssl_cert_verify_cb (int preverify_ok,
         return 1; /* verification successful */
 }
 
+#if OPENSSL_VERSION_NUMBER >= 0x1000200fL && !defined(LIBRESSL_VERSION_NUMBER)
+static int rd_kafka_transport_ssl_cert_fetch_cb(SSL *ssl, void *arg) {
+        rd_kafka_transport_t *rktrans = rd_kafka_curr_transport;
+        rd_kafka_broker_t *rkb;
+        rd_kafka_t *rk;
+        int r;
+        int i;
+        int ret;
+        int num_chain_certs;
+
+        rd_assert(rktrans != NULL);
+        rkb = rktrans->rktrans_rkb;
+        rk = rkb->rkb_rk;
+
+        rd_kafka_cert_fetch_cb_res_t cb_res;
+        size_t buf_size = 0; // start with zero to probe callback for correct size
+        char *buf = NULL;
+        char *leaf_cert_ptr;
+        char *pkey_ptr;
+        char *chain_cert_ptrs[16] = { 0 };
+        size_t leaf_cert_size;
+        size_t pkey_size;
+        size_t chain_cert_sizes[16] = { 0 };
+        rd_kafka_cert_enc_t format;
+
+        char errstr[512] = { 0 };
+        rd_kafka_cert_t *leaf_cert = NULL;
+        rd_kafka_cert_t *pkey = NULL;
+
+        do {
+                if (buf) rd_kafka_mem_free(rk, buf);
+                buf = rd_calloc(1, buf_size);
+                cb_res = rk->rk_conf.ssl.cert_fetch_cb(rk,rkb->rkb_nodename, rkb->rkb_nodeid,
+                                                       buf, &buf_size,
+                                                       &leaf_cert_ptr, &leaf_cert_size,
+                                                       &pkey_ptr, &pkey_size,
+                                                       chain_cert_ptrs, chain_cert_sizes,
+                                                       &format,
+                                                       rk->rk_conf.opaque);
+        } while (cb_res == RD_KAFKA_CERT_FETCH_MORE_BUFFER);
+
+        if (cb_res == RD_KAFKA_CERT_FETCH_NONE) {
+                ret = 1;
+                goto out;
+        } else if (cb_res == RD_KAFKA_CERT_FETCH_ERR) {
```

Comment:
I think this should be logged, and that we should pass errstr+errstr_size to the callback .